budgetkey_data_pipelines/pipelines/budget/national/changes/processed/extract-change-groups.py

```diff
@@ -0,0 +1,237 @@
+        for date_kind, date_changes in itertools.groupby(changes, get_date):
+            date_reserve = [
+                c for c in date_changes
+                if c['budget_code'].startswith('0047') and c['leading_item'] != 47
+                if any(c[field] for field in value_fields)
+            ]
+            logging.debug('Reserve date: kind:%s num:%s' % (
+                date_kind, len(date_reserve)
+            ))
+            num_found = 0  # these two - only to show progress
+            i = 0          #
+            for comb_size in range(2, min(len(date_reserve) + 1, 7)):
+                done = False
+                while not done:
+                    not_selected = list(
+                        x for x in date_reserve
+                        if x['trcode'] not in selected_transfer_codes
+                    )
+                    logging.debug('len(not_selected)=%d' % len(not_selected))
+                    date_transactions = itertools.combinations(
+                        not_selected, comb_size)
+                    found = None
+                    done = True
+                    try:
+                        while True:
+                            i += 1
+                            if i % 100000 == 0:
+                                logging.debug('%s %s %s' % (
+                                    len(date_reserve), num_found, i
+                                ))
+                            transaction = date_transactions.send(found)
+                            found = False
+                            sumvec = sum(
+                                sum(c.get(x, 0) for c in transaction) ** 2
+                                for x in value_fields
+                            )
+                            if sumvec == 0:
+                                transfer_codes = set(
+                                    x['trcode']
+                                    for x in transaction
+                                )
+                                if len(selected_transfer_codes & transfer_codes) > 0:
+                                    continue
+                                selected_transfer_codes.update(transfer_codes)
+                                transfer_codes = sorted(transfer_codes)
+                                num_found += len(transfer_codes)
+                                yield set(transfer_codes),
+                                found = True
+                    except StopIteration:
+                        pass
+        for change in changes:
+            if not change['trcode'] in selected_transfer_codes:
+                selected_transfer_codes.add(change['trcode'])
+                yield {change['trcode']}
+
+    def assign_transactions(groups, changes):
+
+        # This allows to perform search ~10 times faster
+        changes_by_year = {}
+        for x in changes:
+            year = x['year']
+            changes_by_year[year] = changes_by_year.get(year, [])
```

Comment:
the more Pythonic idiom for this would be:
`changes_by_year.setdefault(year, []).append(x)`